core/lib/dal/src/consensus_dal.rs

```diff
@@ -370,13 +396,81 @@ impl ConsensusDal<'_, '_> {
             header.number.0 as i64,
             zksync_protobuf::serde::serialize(cert, serde_json::value::Serializer).unwrap(),
         )
-        .instrument("insert_certificate")
+        .instrument("insert_block_certificate")
+        .report_latency()
+        .execute(&mut txn)
+        .await?;
+        txn.commit().await.context("commit")?;
+        Ok(())
+    }
+
+    /// Inserts a certificate for the L1 batch.
+    ///
+    /// Insertion is allowed even if it creates gaps in the L1 batch history.
+    ///
+    /// It fails if the batch payload is missing or it's not consistent with the QC.
+    pub async fn insert_batch_certificate(
+        &mut self,
+        cert: &attester::BatchQC,
+    ) -> Result<(), InsertCertificateError> {
+        use InsertCertificateError as E;
+        let mut txn = self.storage.start_transaction().await?;
+
+        let _l1_batch_header = txn
+            .blocks_dal()
+            .get_l1_batch_header(zksync_types::L1BatchNumber(cert.message.number.0 as u32))
+            .await?
+            .ok_or(E::MissingPayload)?;
+
+        // TODO: Verify that the certificate matches the stored batch:
```

Comment:
Ah okay, thanks, I'll leave a note about this. I was going to follow the [insert_block_certificate](https://github.com/matter-labs/zksync-era/blob/9673bb14ecabc36d77a2bfa4765e9abe33d04650/core/lib/dal/src/consensus_dal.rs#L386-L388) pattern here. It would be a simple enough sanity check, but the other case of validating `SyncBatch::proof` requires API clients that mean it's definitely not something I'd put in here.

Might be worth considering some kind of wrapper to indicate that validation has happened already.